SmartDeviceLink/SDLHapticManager.m

```diff
@@ -0,0 +1,161 @@
+//
+//  SDLHapticManager.m
+//  SmartDeviceLink-iOS
+//
+//  Copyright © 2017 smartdevicelink. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+#import "SDLHapticManager.h"
+#import "SDLNotificationConstants.h"
+#import "SDLRectangle.h"
+#import "SDLHapticRect.h"
+#import "SDLSendHapticData.h"
+#import "SDLTouch.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SDLHapticManager()
+
+/**
+ The projection window associated with the Haptic Manager
+ */
+@property (nonatomic, strong) UIWindow *projectionWindow;
+
+/**
+ Array of focusable view objects extracted from the projection window
+ */
+@property (nonatomic, strong) NSMutableArray<UIView *> *focusableViews;
+
+/**
+ reference to SDLConnectionManager
+ */
+@property (nonatomic, strong) id<SDLConnectionManagerType> connectionManager;
+@end
+
+
+@implementation SDLHapticManager
+
+- (instancetype)initWithWindow:(UIWindow *)window connectionManager:(id<SDLConnectionManagerType>)connectionManager{
+    self = [super init];
+    if(!self) {
+        return nil;
+    }
+    
+    _projectionWindow = window;
+    _connectionManager = connectionManager;
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdl_projectionViewUpdated:) name:SDLDidProjectionViewUpdate object:nil];
+    
+    return self;
+}
+
+- (void)updateInterfaceLayout {
+    self.focusableViews = [[NSMutableArray alloc] init];
+    [self sdl_parseViewHierarchy:self.projectionWindow.subviews.lastObject];
+    
+    // If there is a preferred view bring that into top of the array
+    NSUInteger preferredViewIndex = [self.focusableViews indexOfObject:self.projectionWindow.subviews.lastObject.preferredFocusedView];
+    if (preferredViewIndex != NSNotFound && self.focusableViews.count > 1) {
+        [self.focusableViews exchangeObjectAtIndex:preferredViewIndex withObjectAtIndex:0];
+    }
+    
+    [self sdl_sendHapticRPC];
+}
+
+/**
+ Crawls through the views recursively and adds focusable view into the member array
+
+ @param currentView is the view hierarchy to be processed
+ */
+- (void)sdl_parseViewHierarchy:(UIView *)currentView {
+    if (currentView == nil) {
+        SDLLogW(@"Error: Cannot parse nil view");
+        return;
+    }
+    
+    NSArray *focusableSubviews = [currentView.subviews filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(UIView *  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
+        return evaluatedObject.canBecomeFocused;
+    }]];
+    
+    //if current view is focusable and it doesn't have any focusable sub views then add the cuurent view and return
+    if (currentView.canBecomeFocused && focusableSubviews.count == 0) {
+        [self.focusableViews addObject:currentView];
+        return;
+    }
+    // if current view has focusable sub views parse them recursively
```

Comment:
So only the focusable subviews get added? The top level focusable view does not? Is that how we're expecting that to work? There should be a test for this code as well.

Reply:
The idea here is that the lowest focusable "leaf" child elements in the view hierarchy should be added as a haptic rects.  The top level focusable view would not be added if it has focusable children.  We added a unit test to cover this case.